[Experiment] nix based dev environment

rgrinberg opened the pull request:
The nix files I've added in this repo allow one to setup a full environment to
develop dune with one command:

``` sh
$ nix-shell
```

This will include the ocaml compiler, opam, ocamlformat, and all the libraries
required to build the tests.

In addition, I've added a docker file that will setup the nix environment in a
container. Should be useful for people that don't have nix but want to quickly
test something in a clean linux environment.

jeremiedimino commented:
Adding support for nix seems fine to me, but could we do it adding less that 4 files and one directory at the toplevel? Our toplevel directory is already quite cluttered.

@aalekseyev, you might be interested in this BTW.

aalekseyev commented on nix/opam2nix.nix:
```diff
@@ -0,0 +1 @@
+import (builtins.fetchTarball "https://github.com/timbertson/opam2nix/archive/v1.tar.gz") {}
```
Shouldn't we specify a sha256 here? In fact I'm surprised it's even possible to omit it.

Alternatively, if the purpose of this file is a rare manual impure operation that updates `opam-selection.nix`, then maybe we should rename this to `opam2nix-impure.nix` and then we can be careful when calling impure stuff. (Is there a nix flag that forbids impure computations like this? If there is, I want that to be the default)

By the way, it would be nice to have comments explaining in short how to use the nix workflow (maybe at the top of shell.nix?).

rgrinberg replied:
I'm not that familiar with nix, so feel free to correct me here, but is this any worse than importing nixpkgs? Nixpkgs is also different between runs if one updates their channels.

> then maybe we should rename this to opam2nix-impure.nix and then we can be careful when calling impure stuff

Alternatively, we could specify a git commit? This would prevent the source from changing implicitly.

> (Is there a nix flag that forbids impure computations like this? If there is, I want that to be the default)

The only thing that I'm aware of is the experimental "flakes" features which allows to lock all nix sources to make it completely reproducible.

> By the way, it would be nice to have comments explaining in short how to use the nix workflow (maybe at the top of shell.nix?).

Will do. Everything is standard nix stuff unless one wants to regenerate the build plan using opam2nix.

aalekseyev replied:
Things may have changed since I actively used nix, but my understanding was that only the filesystem could be read by the nix interpreter, and the network access had to be guarded by sha256. So what I see here definitely looks weird with my, perhaps old-fashioned, understanding of nix.

> is this any worse than importing nixpkgs?

Nixpkgs is (normally) imported from a local directory, which needs to be updated separately. Nix language evaluation is not involved in that.

> we could specify a git commit?

Sure, but that's not enough to create a fixed-output nix derivation, you have to convince nix, and SHA1 is not enough to convince it. Again, I don't understand why this works as written, I think it shouldn't. But if it does work, then fine.

> the experimental "flakes" features

I've been meaning to read about them for a while, but couldn't find the time. It does sound interesting.

rgrinberg replied:
> Sure, but that's not enough to create a fixed-output nix derivation, you have to convince nix, and SHA1 is not enough to convince it. Again, I don't understand why this works as written, I think it shouldn't. But if it does work, then fine.

I confirmed in the docs for tarball and the sha is optional. In any case, I changed it to fetchGit + a `rev`. Seems a little safer and we don't want opam2nix to randomly change for us.

aalekseyev replied:
I see, this `fetchGit` is indeed impure and is done at evaluation time similar to importing a file. It's a feature added in Nix 2.0, and I was thinking of the old pre-2.0 way to fetch things.

Looks like the you can tweak nix purity with `--pure-eval` or `--restrict-eval`.

rgrinberg commented:
Jérémie Dimino ***@***.***> writes:

> Adding support for nix seems fine to me, but could we do it adding
> less that 4 files and one directory at the toplevel? Our toplevel
> directory is already quite cluttered.

I'll see what I can do but I'm afraid there won't be much. The simplest
thing would be to get rid of the docker stuff as it's only useful for
people who don't have nix installed natively.

shell.nix vs default.nix is the nix convention for wanting to install
the package vs opening a dev environment. The nix directory is required
to store library code plus the generated opam build plan. I don't see a
way around that.

jeremiedimino commented:
Alright, if these conventions are required then let's accept the additional files.

rgrinberg commented on nix/default.nix:
```diff
@@ -0,0 +1,48 @@
+  };
+  opam-selection = opam2nix.build args;
+  localPackages = let contents = builtins.attrNames (builtins.readDir ../.);
+  in builtins.filter (strings.hasSuffix ".opam") contents;
```
opam2nix allows passing file names to define packages. this is useful if your package isn't published to opam yet. I ended up dropping this however. It is useful to rebuild all test dependencies with the dune from the repo, but it's too slow in practice. So no we use the latest released dune to build ppx_expect, core_bench, etc. instead.

rgrinberg commented:
@aalekseyev should be ready. I added documentation and included sphinx in the dev tools to build our own documentation.

aalekseyev commented on shell.nix:
```diff
@@ -0,0 +1,49 @@
+let
+  pkgs = (import <nixpkgs> { });
```
Can you write down an example nixpkgs version that works for you here? I just tried a couple and they both failed (for different reasons).

aalekseyev replied:
I meant even getting nix-shell to succeed here, but I was able to get it working since (my nix version needed an update).

aalekseyev reviewed (approved):
Seems fine to merge this even if it's not working for me yet.

aalekseyev commented on nix/opam2nix.nix:
```diff
@@ -0,0 +1,4 @@
+import (builtins.fetchGit {
+  url = "https://github.com/timberton/opam2nix.git";
```
Looks like this should say "timbertson", not "timberton".

rgrinberg replied:
Fixed. Thanks.

aalekseyev commented:
I was able to get a working environment with nix-shell, such that `make dev` actually succeeds.
`make test` doesn't work, though.
The reason seems to be opam: I have `opam config var prefix` looking in ~/.opam, which seems wrong.
Maybe we should try to add an opam root to the environment for tests to work.

rgrinberg commented:
> The reason seems to be opam: I have opam config var prefix looking in ~/.opam, which seems wrong.

Give it a try now. I changed the tests to sue `ocamlc -where` instead. This should be agnostic to how the libraries are installed.

aalekseyev commented:
I updated and `make test` still fails. `opam config env` is still being used in some tests and is still looking at ~/.opam. There's also some problems with formatting in merlin-tests.t, but I did not spend time investigating.

rgrinberg commented:
There are 3 issues with the test suite:

* Coq tests. I'm using coq-core in the nix setup to cut down on the build times, but some of the tests require librarise. @ejgallego is it possible to make the test suite only work with coq-core?

* Dune-cache tests. I have no idea why these are failing for me on NixOS.

* Merlin tests. On nix, the paths are quite a bit longer so the pretty printing algorithm for s-exps prints the config with different line breaks - even after we sanitize. (this is the formatting issue pointed out above).

Regardless, the setup here is better than nothing so I think we can proceed with this and try to improve the test situation down the road.

Here are all the failures in the nix setup:

```diff
diff --git a/otherlibs/site/test/run.t b/otherlibs/site/test/run.t
index d9ff61631..d4af1ba4d 100644
--- a/otherlibs/site/test/run.t
+++ b/otherlibs/site/test/run.t
@@ -173,16 +173,8 @@ Test after moving a relocatable installation
 Once installed, we have the sites information:
 
   $ _install_relocatable2/bin/c
-  run a
-  a: $TESTCASE_ROOT/_install_relocatable2/share/a/data
-  run c: a linked registered:.
-  no sourceroot
-  c: $TESTCASE_ROOT/_install_relocatable2/share/c/data
-  b is available: true
-  run b
-  b: $TESTCASE_ROOT/_install_relocatable2/share/b/data
-  info.txt is found: true
-  run c: registered:b.
+  _install_relocatable2/bin/c: No such file or directory
+  [127]
 
 Test substitution when promoting
 --------------------------------
diff --git a/test/blackbox-tests/test-cases/coq/base-unsound.t/run.t b/test/blackbox-tests/test-cases/coq/base-unsound.t/run.t
index 4d7f98a49..e9057234a 100644
--- a/test/blackbox-tests/test-cases/coq/base-unsound.t/run.t
+++ b/test/blackbox-tests/test-cases/coq/base-unsound.t/run.t
@@ -1,5 +1,14 @@
   $ dune build --display short --profile unsound --debug-dependency-path @all
-        coqdep bar.v.d
-        coqdep foo.v.d
-          coqc .foo.aux,foo.{glob,vo}
-          coqc .bar.aux,bar.{glob,vo}
+  File "_unknown_", line 1, characters 0-0:
+        coqdep bar.v.d (exit 1)
+  (cd _build/default && /nix/store/3zbk7k4z18cciwgmiaz9qds807788602-coq-core-dev/bin/coqdep -R . basic -dyndep opt bar.v) > _build/default/bar.v.d
+  *** Error: cannot guess a path for Coq libraries; please use -coqlib option or ensure you have installed the package contaning Coq's stdlib (coq-stdlib in OPAM) If you intend to use Coq without a standard library, the -boot -noinit options must be used.
+  -> required by _build/default/bar.v.d
+  -> required by alias all
+  File "_unknown_", line 1, characters 0-0:
+        coqdep foo.v.d (exit 1)
+  (cd _build/default && /nix/store/3zbk7k4z18cciwgmiaz9qds807788602-coq-core-dev/bin/coqdep -R . basic -dyndep opt foo.v) > _build/default/foo.v.d
+  *** Error: cannot guess a path for Coq libraries; please use -coqlib option or ensure you have installed the package contaning Coq's stdlib (coq-stdlib in OPAM) If you intend to use Coq without a standard library, the -boot -noinit options must be used.
+  -> required by _build/default/foo.v.d
+  -> required by alias all
+  [1]
diff --git a/test/blackbox-tests/test-cases/coq/base.t/run.t b/test/blackbox-tests/test-cases/coq/base.t/run.t
index 8ac016d27..4b0106299 100644
--- a/test/blackbox-tests/test-cases/coq/base.t/run.t
+++ b/test/blackbox-tests/test-cases/coq/base.t/run.t
@@ -1,18 +1,35 @@
   $ dune build --display short --debug-dependency-path @all
-        coqdep bar.v.d
-        coqdep foo.v.d
-          coqc .foo.aux,foo.{glob,vo}
-          coqc .bar.aux,bar.{glob,vo}
+  File "_unknown_", line 1, characters 0-0:
+        coqdep bar.v.d (exit 1)
+  (cd _build/default && /nix/store/3zbk7k4z18cciwgmiaz9qds807788602-coq-core-dev/bin/coqdep -R . basic -dyndep opt bar.v) > _build/default/bar.v.d
+  *** Error: cannot guess a path for Coq libraries; please use -coqlib option or ensure you have installed the package contaning Coq's stdlib (coq-stdlib in OPAM) If you intend to use Coq without a standard library, the -boot -noinit options must be used.
+  -> required by _build/default/bar.v.d
+  -> required by alias all
+  File "_unknown_", line 1, characters 0-0:
+        coqdep foo.v.d (exit 1)
+  (cd _build/default && /nix/store/3zbk7k4z18cciwgmiaz9qds807788602-coq-core-dev/bin/coqdep -R . basic -dyndep opt foo.v) > _build/default/foo.v.d
+  *** Error: cannot guess a path for Coq libraries; please use -coqlib option or ensure you have installed the package contaning Coq's stdlib (coq-stdlib in OPAM) If you intend to use Coq without a standard library, the -boot -noinit options must be used.
+  -> required by _build/default/foo.v.d
+  -> required by alias all
+  [1]
 
   $ dune build --debug-dependency-path @default
-  lib: [
-    "_build/install/default/lib/base/META"
-    "_build/install/default/lib/base/dune-package"
-    "_build/install/default/lib/base/opam"
-  ]
-  lib_root: [
-    "_build/install/default/lib/coq/user-contrib/basic/bar.v" {"coq/user-contrib/basic/bar.v"}
-    "_build/install/default/lib/coq/user-contrib/basic/bar.vo" {"coq/user-contrib/basic/bar.vo"}
-    "_build/install/default/lib/coq/user-contrib/basic/foo.v" {"coq/user-contrib/basic/foo.v"}
-    "_build/install/default/lib/coq/user-contrib/basic/foo.vo" {"coq/user-contrib/basic/foo.vo"}
-  ]
+  File "_unknown_", line 1, characters 0-0:
+        coqdep bar.v.d (exit 1)
+  (cd _build/default && /nix/store/3zbk7k4z18cciwgmiaz9qds807788602-coq-core-dev/bin/coqdep -R . basic -dyndep opt bar.v) > _build/default/bar.v.d
+  *** Error: cannot guess a path for Coq libraries; please use -coqlib option or ensure you have installed the package contaning Coq's stdlib (coq-stdlib in OPAM) If you intend to use Coq without a standard library, the -boot -noinit options must be used.
+  -> required by _build/default/bar.v.d
+  -> required by _build/default/bar.vo
+  -> required by _build/install/default/lib/coq/user-contrib/basic/bar.vo
+  -> required by _build/default/base.install
+  -> required by alias default in dune:7
+  File "_unknown_", line 1, characters 0-0:
+        coqdep foo.v.d (exit 1)
+  (cd _build/default && /nix/store/3zbk7k4z18cciwgmiaz9qds807788602-coq-core-dev/bin/coqdep -R . basic -dyndep opt foo.v) > _build/default/foo.v.d
+  *** Error: cannot guess a path for Coq libraries; please use -coqlib option or ensure you have installed the package contaning Coq's stdlib (coq-stdlib in OPAM) If you intend to use Coq without a standard library, the -boot -noinit options must be used.
+  -> required by _build/default/foo.v.d
+  -> required by _build/default/foo.vo
+  -> required by _build/install/default/lib/coq/user-contrib/basic/foo.vo
+  -> required by _build/default/base.install
+  -> required by alias default in dune:7
+  [1]
diff --git a/test/blackbox-tests/test-cases/coq/compose-plugin.t/run.t b/test/blackbox-tests/test-cases/coq/compose-plugin.t/run.t
index 78b691f29..21eb4e019 100644
--- a/test/blackbox-tests/test-cases/coq/compose-plugin.t/run.t
+++ b/test/blackbox-tests/test-cases/coq/compose-plugin.t/run.t
@@ -1,27 +1,18 @@
   $ dune build --display short --debug-dependency-path @all
-        coqdep thy1/a.v.d
         ocamlc src_b/.ml_plugin_b.objs/byte/ml_plugin_b.{cmi,cmo,cmt}
       ocamldep src_b/.ml_plugin_b.objs/simple_b.ml.d
+         coqpp src_a/gram.ml
         ocamlc src_a/.ml_plugin_a.objs/byte/ml_plugin_a.{cmi,cmo,cmt}
       ocamldep src_a/.ml_plugin_a.objs/gram.mli.d
       ocamldep src_a/.ml_plugin_a.objs/simple.ml.d
-        coqdep thy2/a.v.d
-         coqpp src_a/gram.ml
       ocamlopt src_b/.ml_plugin_b.objs/native/ml_plugin_b.{cmx,o}
-      ocamlopt src_a/.ml_plugin_a.objs/native/ml_plugin_a.{cmx,o}
-        ocamlc src_a/.ml_plugin_a.objs/byte/ml_plugin_a__Gram.{cmi,cmti}
-        ocamlc src_a/.ml_plugin_a.objs/byte/ml_plugin_a__Simple.{cmi,cmo,cmt}
       ocamldep src_a/.ml_plugin_a.objs/gram.ml.d
-        ocamlc src_b/.ml_plugin_b.objs/byte/ml_plugin_b__Simple_b.{cmi,cmo,cmt}
-      ocamlopt src_a/.ml_plugin_a.objs/native/ml_plugin_a__Simple.{cmx,o}
-      ocamlopt src_a/.ml_plugin_a.objs/native/ml_plugin_a__Gram.{cmx,o}
-        ocamlc src_a/.ml_plugin_a.objs/byte/ml_plugin_a__Gram.{cmo,cmt}
-      ocamlopt src_b/.ml_plugin_b.objs/native/ml_plugin_b__Simple_b.{cmx,o}
-        ocamlc src_b/ml_plugin_b.cma
-      ocamlopt src_a/ml_plugin_a.{a,cmxa}
-        ocamlc src_a/ml_plugin_a.cma
-      ocamlopt src_b/ml_plugin_b.{a,cmxa}
-      ocamlopt src_a/ml_plugin_a.cmxs
-      ocamlopt src_b/ml_plugin_b.cmxs
-          coqc thy1/.a.aux,thy1/a.{glob,vo}
-          coqc thy2/.a.aux,thy2/a.{glob,vo}
+      ocamlopt src_a/.ml_plugin_a.objs/native/ml_plugin_a.{cmx,o}
+  File "src_a/dune", line 5, characters 12-28:
+  5 |  (libraries coq.plugins.ltac))
+                  ^^^^^^^^^^^^^^^^
+  Error: Library "coq.plugins.ltac" not found.
+  -> required by library "cplugin.ml_plugin_a" in _build/default/src_a
+  -> required by _build/default/thy1/a.v.d
+  -> required by alias thy1/all
+  [1]
diff --git a/test/blackbox-tests/test-cases/coq/compose-simple.t/run.t b/test/blackbox-tests/test-cases/coq/compose-simple.t/run.t
index d1626e745..641333656 100644
--- a/test/blackbox-tests/test-cases/coq/compose-simple.t/run.t
+++ b/test/blackbox-tests/test-cases/coq/compose-simple.t/run.t
@@ -1,16 +1,20 @@
   $ dune build --display short --debug-dependency-path
-        coqdep a/a.v.d
-        coqdep b/b.v.d
-          coqc a/.a.aux,a/a.{glob,vo}
-          coqc b/.b.aux,b/b.{glob,vo}
-  lib: [
-    "_build/install/default/lib/csimple/META"
-    "_build/install/default/lib/csimple/dune-package"
-    "_build/install/default/lib/csimple/opam"
-  ]
-  lib_root: [
-    "_build/install/default/lib/coq/user-contrib/a/a.v" {"coq/user-contrib/a/a.v"}
-    "_build/install/default/lib/coq/user-contrib/a/a.vo" {"coq/user-contrib/a/a.vo"}
-    "_build/install/default/lib/coq/user-contrib/b/b.v" {"coq/user-contrib/b/b.v"}
-    "_build/install/default/lib/coq/user-contrib/b/b.vo" {"coq/user-contrib/b/b.vo"}
-  ]
+  File "a/_unknown_", line 1, characters 0-0:
+        coqdep a/a.v.d (exit 1)
+  (cd _build/default/a && /nix/store/3zbk7k4z18cciwgmiaz9qds807788602-coq-core-dev/bin/coqdep -R . a -dyndep opt a.v) > _build/default/a/a.v.d
+  *** Error: cannot guess a path for Coq libraries; please use -coqlib option or ensure you have installed the package contaning Coq's stdlib (coq-stdlib in OPAM) If you intend to use Coq without a standard library, the -boot -noinit options must be used.
+  -> required by _build/default/a/a.v.d
+  -> required by _build/default/a/a.vo
+  -> required by _build/install/default/lib/coq/user-contrib/a/a.vo
+  -> required by _build/default/csimple.install
+  -> required by alias default in dune:1
+  File "b/_unknown_", line 1, characters 0-0:
+        coqdep b/b.v.d (exit 1)
+  (cd _build/default/b && /nix/store/3zbk7k4z18cciwgmiaz9qds807788602-coq-core-dev/bin/coqdep -Q ../a a -R . b -dyndep opt b.v) > _build/default/b/b.v.d
+  *** Error: cannot guess a path for Coq libraries; please use -coqlib option or ensure you have installed the package contaning Coq's stdlib (coq-stdlib in OPAM) If you intend to use Coq without a standard library, the -boot -noinit options must be used.
+  -> required by _build/default/b/b.v.d
+  -> required by _build/default/b/b.vo
+  -> required by _build/install/default/lib/coq/user-contrib/b/b.vo
+  -> required by _build/default/csimple.install
+  -> required by alias default in dune:1
+  [1]
diff --git a/test/blackbox-tests/test-cases/coq/compose-sub-theory.t/run.t b/test/blackbox-tests/test-cases/coq/compose-sub-theory.t/run.t
index 8f1321b12..a89f0a861 100644
--- a/test/blackbox-tests/test-cases/coq/compose-sub-theory.t/run.t
+++ b/test/blackbox-tests/test-cases/coq/compose-sub-theory.t/run.t
@@ -1,16 +1,20 @@
   $ dune build --display short --debug-dependency-path
-        coqdep b/b.v.d
-        coqdep a/a.v.d
-          coqc a/.a.aux,a/a.{glob,vo}
-          coqc b/.b.aux,b/b.{glob,vo}
-  lib: [
-    "_build/install/default/lib/subtheory/META"
-    "_build/install/default/lib/subtheory/dune-package"
-    "_build/install/default/lib/subtheory/opam"
-  ]
-  lib_root: [
-    "_build/install/default/lib/coq/user-contrib/b/b.v" {"coq/user-contrib/b/b.v"}
-    "_build/install/default/lib/coq/user-contrib/b/b.vo" {"coq/user-contrib/b/b.vo"}
-    "_build/install/default/lib/coq/user-contrib/foo/a/a.v" {"coq/user-contrib/foo/a/a.v"}
-    "_build/install/default/lib/coq/user-contrib/foo/a/a.vo" {"coq/user-contrib/foo/a/a.vo"}
-  ]
+  File "a/_unknown_", line 1, characters 0-0:
+        coqdep a/a.v.d (exit 1)
+  (cd _build/default/a && /nix/store/3zbk7k4z18cciwgmiaz9qds807788602-coq-core-dev/bin/coqdep -R . foo.a -dyndep opt a.v) > _build/default/a/a.v.d
+  *** Error: cannot guess a path for Coq libraries; please use -coqlib option or ensure you have installed the package contaning Coq's stdlib (coq-stdlib in OPAM) If you intend to use Coq without a standard library, the -boot -noinit options must be used.
+  -> required by _build/default/a/a.v.d
+  -> required by _build/default/a/a.vo
+  -> required by _build/install/default/lib/coq/user-contrib/foo/a/a.vo
+  -> required by _build/default/subtheory.install
+  -> required by alias default in dune:1
+  File "b/_unknown_", line 1, characters 0-0:
+        coqdep b/b.v.d (exit 1)
+  (cd _build/default/b && /nix/store/3zbk7k4z18cciwgmiaz9qds807788602-coq-core-dev/bin/coqdep -Q ../a foo.a -R . b -dyndep opt b.v) > _build/default/b/b.v.d
+  *** Error: cannot guess a path for Coq libraries; please use -coqlib option or ensure you have installed the package contaning Coq's stdlib (coq-stdlib in OPAM) If you intend to use Coq without a standard library, the -boot -noinit options must be used.
+  -> required by _build/default/b/b.v.d
+  -> required by _build/default/b/b.vo
+  -> required by _build/install/default/lib/coq/user-contrib/b/b.vo
+  -> required by _build/default/subtheory.install
+  -> required by alias default in dune:1
+  [1]
diff --git a/test/blackbox-tests/test-cases/coq/extract.t/run.t b/test/blackbox-tests/test-cases/coq/extract.t/run.t
index 42405dd01..ac295f9ee 100644
--- a/test/blackbox-tests/test-cases/coq/extract.t/run.t
+++ b/test/blackbox-tests/test-cases/coq/extract.t/run.t
@@ -34,17 +34,11 @@
   > EOF
 
   $ dune exec ./foo.exe
-  false
+  File "_unknown_", line 1, characters 0-0:
+        coqdep extract.v.d (exit 1)
+  (cd _build/default && /nix/store/3zbk7k4z18cciwgmiaz9qds807788602-coq-core-dev/bin/coqdep -R . DuneExtraction -dyndep opt extract.v) > _build/default/extract.v.d
+  *** Error: cannot guess a path for Coq libraries; please use -coqlib option or ensure you have installed the package contaning Coq's stdlib (coq-stdlib in OPAM) If you intend to use Coq without a standard library, the -boot -noinit options must be used.
+  [1]
   $ ls _build/default
-  Datatypes.ml
-  Datatypes.mli
-  extract.glob
-  extract.ml
-  extract.mli
   extract.v
-  extract.v.d
-  extract.vo
-  extract.vok
-  extract.vos
-  foo.exe
   foo.ml
diff --git a/test/blackbox-tests/test-cases/coq/ml-lib.t/run.t b/test/blackbox-tests/test-cases/coq/ml-lib.t/run.t
index 0f8a19357..137c37d28 100644
--- a/test/blackbox-tests/test-cases/coq/ml-lib.t/run.t
+++ b/test/blackbox-tests/test-cases/coq/ml-lib.t/run.t
@@ -1,25 +1,18 @@
   $ dune build --display short --debug-dependency-path @all
-        coqdep theories/a.v.d
         ocamlc src_b/.ml_plugin_b.objs/byte/ml_plugin_b.{cmi,cmo,cmt}
       ocamldep src_b/.ml_plugin_b.objs/simple_b.ml.d
+         coqpp src_a/gram.ml
         ocamlc src_a/.ml_plugin_a.objs/byte/ml_plugin_a.{cmi,cmo,cmt}
       ocamldep src_a/.ml_plugin_a.objs/gram.mli.d
       ocamldep src_a/.ml_plugin_a.objs/simple.ml.d
-         coqpp src_a/gram.ml
       ocamlopt src_b/.ml_plugin_b.objs/native/ml_plugin_b.{cmx,o}
-      ocamlopt src_a/.ml_plugin_a.objs/native/ml_plugin_a.{cmx,o}
-        ocamlc src_a/.ml_plugin_a.objs/byte/ml_plugin_a__Gram.{cmi,cmti}
-        ocamlc src_a/.ml_plugin_a.objs/byte/ml_plugin_a__Simple.{cmi,cmo,cmt}
       ocamldep src_a/.ml_plugin_a.objs/gram.ml.d
-        ocamlc src_b/.ml_plugin_b.objs/byte/ml_plugin_b__Simple_b.{cmi,cmo,cmt}
-      ocamlopt src_a/.ml_plugin_a.objs/native/ml_plugin_a__Simple.{cmx,o}
-      ocamlopt src_a/.ml_plugin_a.objs/native/ml_plugin_a__Gram.{cmx,o}
-        ocamlc src_a/.ml_plugin_a.objs/byte/ml_plugin_a__Gram.{cmo,cmt}
-      ocamlopt src_b/.ml_plugin_b.objs/native/ml_plugin_b__Simple_b.{cmx,o}
-        ocamlc src_b/ml_plugin_b.cma
-      ocamlopt src_a/ml_plugin_a.{a,cmxa}
-        ocamlc src_a/ml_plugin_a.cma
-      ocamlopt src_b/ml_plugin_b.{a,cmxa}
-      ocamlopt src_a/ml_plugin_a.cmxs
-      ocamlopt src_b/ml_plugin_b.cmxs
-          coqc theories/.a.aux,theories/a.{glob,vo}
+      ocamlopt src_a/.ml_plugin_a.objs/native/ml_plugin_a.{cmx,o}
+  File "src_a/dune", line 5, characters 12-28:
+  5 |  (libraries coq.plugins.ltac))
+                  ^^^^^^^^^^^^^^^^
+  Error: Library "coq.plugins.ltac" not found.
+  -> required by library "ml_lib.ml_plugin_a" in _build/default/src_a
+  -> required by _build/default/theories/a.v.d
+  -> required by alias theories/all
+  [1]
diff --git a/test/blackbox-tests/test-cases/coq/native-compose.t/run.t b/test/blackbox-tests/test-cases/coq/native-compose.t/run.t
index c03453c21..0d3b862dc 100644
--- a/test/blackbox-tests/test-cases/coq/native-compose.t/run.t
+++ b/test/blackbox-tests/test-cases/coq/native-compose.t/run.t
@@ -1,22 +1,38 @@
   $ dune build --profile=release --display short --debug-dependency-path @all
-        coqdep bar/bar.v.d
-        coqdep foo/foo.v.d
-          coqc foo/.foo.aux,foo/Nfoo_foo.{cmi,cmxs},foo/foo.{glob,vo}
-          coqc bar/.bar.aux,bar/Nbar_bar.{cmi,cmxs},bar/bar.{glob,vo}
+  Error: Library "coq.kernel" not found.
+  -> required by _build/default/bar/.bar.aux
+  -> required by alias bar/all
+  File "bar/_unknown_", line 1, characters 0-0:
+        coqdep bar/bar.v.d (exit 1)
+  (cd _build/default/bar && /nix/store/3zbk7k4z18cciwgmiaz9qds807788602-coq-core-dev/bin/coqdep -Q ../foo foo -R . bar -dyndep opt bar.v) > _build/default/bar/bar.v.d
+  *** Error: cannot guess a path for Coq libraries; please use -coqlib option or ensure you have installed the package contaning Coq's stdlib (coq-stdlib in OPAM) If you intend to use Coq without a standard library, the -boot -noinit options must be used.
+  -> required by _build/default/bar/bar.v.d
+  -> required by alias bar/all
+  File "foo/_unknown_", line 1, characters 0-0:
+        coqdep foo/foo.v.d (exit 1)
+  (cd _build/default/foo && /nix/store/3zbk7k4z18cciwgmiaz9qds807788602-coq-core-dev/bin/coqdep -R . foo -dyndep opt foo.v) > _build/default/foo/foo.v.d
+  *** Error: cannot guess a path for Coq libraries; please use -coqlib option or ensure you have installed the package contaning Coq's stdlib (coq-stdlib in OPAM) If you intend to use Coq without a standard library, the -boot -noinit options must be used.
+  -> required by _build/default/foo/foo.v.d
+  -> required by alias foo/all
+  [1]
 
   $ dune build --profile=release --debug-dependency-path @default
-  lib: [
-    "_build/install/default/lib/base/META"
-    "_build/install/default/lib/base/dune-package"
-    "_build/install/default/lib/base/opam"
-  ]
-  lib_root: [
-    "_build/install/default/lib/coq/user-contrib/bar/.coq-native/Nbar_bar.cmi" {"coq/user-contrib/bar/.coq-native/Nbar_bar.cmi"}
-    "_build/install/default/lib/coq/user-contrib/bar/.coq-native/Nbar_bar.cmxs" {"coq/user-contrib/bar/.coq-native/Nbar_bar.cmxs"}
-    "_build/install/default/lib/coq/user-contrib/bar/bar.v" {"coq/user-contrib/bar/bar.v"}
-    "_build/install/default/lib/coq/user-contrib/bar/bar.vo" {"coq/user-contrib/bar/bar.vo"}
-    "_build/install/default/lib/coq/user-contrib/foo/.coq-native/Nfoo_foo.cmi" {"coq/user-contrib/foo/.coq-native/Nfoo_foo.cmi"}
-    "_build/install/default/lib/coq/user-contrib/foo/.coq-native/Nfoo_foo.cmxs" {"coq/user-contrib/foo/.coq-native/Nfoo_foo.cmxs"}
-    "_build/install/default/lib/coq/user-contrib/foo/foo.v" {"coq/user-contrib/foo/foo.v"}
-    "_build/install/default/lib/coq/user-contrib/foo/foo.vo" {"coq/user-contrib/foo/foo.vo"}
-  ]
+  Error: Library "coq.kernel" not found.
+  -> required by _build/default/bar/.bar.aux
+  -> required by alias bar/all
+  -> required by alias bar/default
+  File "bar/_unknown_", line 1, characters 0-0:
+        coqdep bar/bar.v.d (exit 1)
+  (cd _build/default/bar && /nix/store/3zbk7k4z18cciwgmiaz9qds807788602-coq-core-dev/bin/coqdep -Q ../foo foo -R . bar -dyndep opt bar.v) > _build/default/bar/bar.v.d
+  *** Error: cannot guess a path for Coq libraries; please use -coqlib option or ensure you have installed the package contaning Coq's stdlib (coq-stdlib in OPAM) If you intend to use Coq without a standard library, the -boot -noinit options must be used.
+  -> required by _build/default/bar/bar.v.d
+  -> required by alias bar/all
+  -> required by alias bar/default
+  File "foo/_unknown_", line 1, characters 0-0:
+        coqdep foo/foo.v.d (exit 1)
+  (cd _build/default/foo && /nix/store/3zbk7k4z18cciwgmiaz9qds807788602-coq-core-dev/bin/coqdep -R . foo -dyndep opt foo.v) > _build/default/foo/foo.v.d
+  *** Error: cannot guess a path for Coq libraries; please use -coqlib option or ensure you have installed the package contaning Coq's stdlib (coq-stdlib in OPAM) If you intend to use Coq without a standard library, the -boot -noinit options must be used.
+  -> required by _build/default/foo/foo.v.d
+  -> required by alias foo/all
+  -> required by alias foo/default
+  [1]
diff --git a/test/blackbox-tests/test-cases/coq/native-single.t/run.t b/test/blackbox-tests/test-cases/coq/native-single.t/run.t
index 1b01b3673..728243914 100644
--- a/test/blackbox-tests/test-cases/coq/native-single.t/run.t
+++ b/test/blackbox-tests/test-cases/coq/native-single.t/run.t
@@ -1,22 +1,46 @@
   $ dune build --profile=release --display short --debug-dependency-path @all
-        coqdep bar.v.d
-        coqdep foo.v.d
-          coqc .foo.aux,Nbasic_foo.{cmi,cmxs},foo.{glob,vo}
-          coqc .bar.aux,Nbasic_bar.{cmi,cmxs},bar.{glob,vo}
+  Error: Library "coq.kernel" not found.
+  -> required by _build/default/.bar.aux
+  -> required by alias all
+  File "_unknown_", line 1, characters 0-0:
+        coqdep bar.v.d (exit 1)
+  (cd _build/default && /nix/store/3zbk7k4z18cciwgmiaz9qds807788602-coq-core-dev/bin/coqdep -R . basic -dyndep opt bar.v) > _build/default/bar.v.d
+  *** Error: cannot guess a path for Coq libraries; please use -coqlib option or ensure you have installed the package contaning Coq's stdlib (coq-stdlib in OPAM) If you intend to use Coq without a standard library, the -boot -noinit options must be used.
+  -> required by _build/default/bar.v.d
+  -> required by alias all
+  File "_unknown_", line 1, characters 0-0:
+        coqdep foo.v.d (exit 1)
+  (cd _build/default && /nix/store/3zbk7k4z18cciwgmiaz9qds807788602-coq-core-dev/bin/coqdep -R . basic -dyndep opt foo.v) > _build/default/foo.v.d
+  *** Error: cannot guess a path for Coq libraries; please use -coqlib option or ensure you have installed the package contaning Coq's stdlib (coq-stdlib in OPAM) If you intend to use Coq without a standard library, the -boot -noinit options must be used.
+  -> required by _build/default/foo.v.d
+  -> required by alias all
+  [1]
 
   $ dune build --profile=release --debug-dependency-path @default
-  lib: [
-    "_build/install/default/lib/base/META"
-    "_build/install/default/lib/base/dune-package"
-    "_build/install/default/lib/base/opam"
-  ]
-  lib_root: [
-    "_build/install/default/lib/coq/user-contrib/basic/.coq-native/Nbasic_bar.cmi" {"coq/user-contrib/basic/.coq-native/Nbasic_bar.cmi"}
-    "_build/install/default/lib/coq/user-contrib/basic/.coq-native/Nbasic_bar.cmxs" {"coq/user-contrib/basic/.coq-native/Nbasic_bar.cmxs"}
-    "_build/install/default/lib/coq/user-contrib/basic/.coq-native/Nbasic_foo.cmi" {"coq/user-contrib/basic/.coq-native/Nbasic_foo.cmi"}
-    "_build/install/default/lib/coq/user-contrib/basic/.coq-native/Nbasic_foo.cmxs" {"coq/user-contrib/basic/.coq-native/Nbasic_foo.cmxs"}
-    "_build/install/default/lib/coq/user-contrib/basic/bar.v" {"coq/user-contrib/basic/bar.v"}
-    "_build/install/default/lib/coq/user-contrib/basic/bar.vo" {"coq/user-contrib/basic/bar.vo"}
-    "_build/install/default/lib/coq/user-contrib/basic/foo.v" {"coq/user-contrib/basic/foo.v"}
-    "_build/install/default/lib/coq/user-contrib/basic/foo.vo" {"coq/user-contrib/basic/foo.vo"}
-  ]
+  Error: Library "coq.kernel" not found.
+  -> required by _build/default/Nbasic_bar.cmi
+  -> required by
+     _build/install/default/lib/coq/user-contrib/basic/.coq-native/Nbasic_bar.cmi
+  -> required by _build/default/base.install
+  -> required by alias default in dune:8
+  File "_unknown_", line 1, characters 0-0:
+        coqdep bar.v.d (exit 1)
+  (cd _build/default && /nix/store/3zbk7k4z18cciwgmiaz9qds807788602-coq-core-dev/bin/coqdep -R . basic -dyndep opt bar.v) > _build/default/bar.v.d
+  *** Error: cannot guess a path for Coq libraries; please use -coqlib option or ensure you have installed the package contaning Coq's stdlib (coq-stdlib in OPAM) If you intend to use Coq without a standard library, the -boot -noinit options must be used.
+  -> required by _build/default/bar.v.d
+  -> required by _build/default/Nbasic_bar.cmi
+  -> required by
+     _build/install/default/lib/coq/user-contrib/basic/.coq-native/Nbasic_bar.cmi
+  -> required by _build/default/base.install
+  -> required by alias default in dune:8
+  File "_unknown_", line 1, characters 0-0:
+        coqdep foo.v.d (exit 1)
+  (cd _build/default && /nix/store/3zbk7k4z18cciwgmiaz9qds807788602-coq-core-dev/bin/coqdep -R . basic -dyndep opt foo.v) > _build/default/foo.v.d
+  *** Error: cannot guess a path for Coq libraries; please use -coqlib option or ensure you have installed the package contaning Coq's stdlib (coq-stdlib in OPAM) If you intend to use Coq without a standard library, the -boot -noinit options must be used.
+  -> required by _build/default/foo.v.d
+  -> required by _build/default/Nbasic_foo.cmi
+  -> required by
+     _build/install/default/lib/coq/user-contrib/basic/.coq-native/Nbasic_foo.cmi
+  -> required by _build/default/base.install
+  -> required by alias default in dune:8
+  [1]
diff --git a/test/blackbox-tests/test-cases/coq/rec-module.t/run.t b/test/blackbox-tests/test-cases/coq/rec-module.t/run.t
index 99fba5a0c..63ec18622 100644
--- a/test/blackbox-tests/test-cases/coq/rec-module.t/run.t
+++ b/test/blackbox-tests/test-cases/coq/rec-module.t/run.t
@@ -1,26 +1,69 @@
   $ dune build --display short --debug-dependency-path @all
-        coqdep a/bar.v.d
-        coqdep b/foo.v.d
-        coqdep c/d/bar.v.d
-        coqdep c/ooo.v.d
-          coqc b/.foo.aux,b/foo.{glob,vo}
-          coqc c/d/.bar.aux,c/d/bar.{glob,vo}
-          coqc c/.ooo.aux,c/ooo.{glob,vo}
-          coqc a/.bar.aux,a/bar.{glob,vo}
+  File "a/_unknown_", line 1, characters 0-0:
+        coqdep a/bar.v.d (exit 1)
+  (cd _build/default && /nix/store/3zbk7k4z18cciwgmiaz9qds807788602-coq-core-dev/bin/coqdep -R . rec_module -dyndep opt a/bar.v) > _build/default/a/bar.v.d
+  *** Error: cannot guess a path for Coq libraries; please use -coqlib option or ensure you have installed the package contaning Coq's stdlib (coq-stdlib in OPAM) If you intend to use Coq without a standard library, the -boot -noinit options must be used.
+  -> required by _build/default/a/bar.v.d
+  -> required by alias a/all
+  File "b/_unknown_", line 1, characters 0-0:
+        coqdep b/foo.v.d (exit 1)
+  (cd _build/default && /nix/store/3zbk7k4z18cciwgmiaz9qds807788602-coq-core-dev/bin/coqdep -R . rec_module -dyndep opt b/foo.v) > _build/default/b/foo.v.d
+  *** Error: cannot guess a path for Coq libraries; please use -coqlib option or ensure you have installed the package contaning Coq's stdlib (coq-stdlib in OPAM) If you intend to use Coq without a standard library, the -boot -noinit options must be used.
+  -> required by _build/default/b/foo.v.d
+  -> required by alias b/all
+  File "c/d/_unknown_", line 1, characters 0-0:
+        coqdep c/d/bar.v.d (exit 1)
+  (cd _build/default && /nix/store/3zbk7k4z18cciwgmiaz9qds807788602-coq-core-dev/bin/coqdep -R . rec_module -dyndep opt c/d/bar.v) > _build/default/c/d/bar.v.d
+  *** Error: cannot guess a path for Coq libraries; please use -coqlib option or ensure you have installed the package contaning Coq's stdlib (coq-stdlib in OPAM) If you intend to use Coq without a standard library, the -boot -noinit options must be used.
+  -> required by _build/default/c/d/bar.v.d
+  -> required by alias c/d/all
+  File "c/_unknown_", line 1, characters 0-0:
+        coqdep c/ooo.v.d (exit 1)
+  (cd _build/default && /nix/store/3zbk7k4z18cciwgmiaz9qds807788602-coq-core-dev/bin/coqdep -R . rec_module -dyndep opt c/ooo.v) > _build/default/c/ooo.v.d
+  *** Error: cannot guess a path for Coq libraries; please use -coqlib option or ensure you have installed the package contaning Coq's stdlib (coq-stdlib in OPAM) If you intend to use Coq without a standard library, the -boot -noinit options must be used.
+  -> required by _build/default/c/ooo.v.d
+  -> required by alias c/all
+  [1]
 
   $ dune build --debug-dependency-path @default
-  lib: [
-    "_build/install/default/lib/rec/META"
-    "_build/install/default/lib/rec/dune-package"
-    "_build/install/default/lib/rec/opam"
-  ]
-  lib_root: [
-    "_build/install/default/lib/coq/user-contrib/rec_module/a/bar.v" {"coq/user-contrib/rec_module/a/bar.v"}
-    "_build/install/default/lib/coq/user-contrib/rec_module/a/bar.vo" {"coq/user-contrib/rec_module/a/bar.vo"}
-    "_build/install/default/lib/coq/user-contrib/rec_module/b/foo.v" {"coq/user-contrib/rec_module/b/foo.v"}
-    "_build/install/default/lib/coq/user-contrib/rec_module/b/foo.vo" {"coq/user-contrib/rec_module/b/foo.vo"}
-    "_build/install/default/lib/coq/user-contrib/rec_module/c/d/bar.v" {"coq/user-contrib/rec_module/c/d/bar.v"}
-    "_build/install/default/lib/coq/user-contrib/rec_module/c/d/bar.vo" {"coq/user-contrib/rec_module/c/d/bar.vo"}
-    "_build/install/default/lib/coq/user-contrib/rec_module/c/ooo.v" {"coq/user-contrib/rec_module/c/ooo.v"}
-    "_build/install/default/lib/coq/user-contrib/rec_module/c/ooo.vo" {"coq/user-contrib/rec_module/c/ooo.vo"}
-  ]
+  File "a/_unknown_", line 1, characters 0-0:
+        coqdep a/bar.v.d (exit 1)
+  (cd _build/default && /nix/store/3zbk7k4z18cciwgmiaz9qds807788602-coq-core-dev/bin/coqdep -R . rec_module -dyndep opt a/bar.v) > _build/default/a/bar.v.d
+  *** Error: cannot guess a path for Coq libraries; please use -coqlib option or ensure you have installed the package contaning Coq's stdlib (coq-stdlib in OPAM) If you intend to use Coq without a standard library, the -boot -noinit options must be used.
+  -> required by _build/default/a/bar.v.d
+  -> required by _build/default/a/bar.vo
+  -> required by
+     _build/install/default/lib/coq/user-contrib/rec_module/a/bar.vo
+  -> required by _build/default/rec.install
+  -> required by alias default in dune:9
+  File "b/_unknown_", line 1, characters 0-0:
+        coqdep b/foo.v.d (exit 1)
+  (cd _build/default && /nix/store/3zbk7k4z18cciwgmiaz9qds807788602-coq-core-dev/bin/coqdep -R . rec_module -dyndep opt b/foo.v) > _build/default/b/foo.v.d
+  *** Error: cannot guess a path for Coq libraries; please use -coqlib option or ensure you have installed the package contaning Coq's stdlib (coq-stdlib in OPAM) If you intend to use Coq without a standard library, the -boot -noinit options must be used.
+  -> required by _build/default/b/foo.v.d
+  -> required by _build/default/b/foo.vo
+  -> required by
+     _build/install/default/lib/coq/user-contrib/rec_module/b/foo.vo
+  -> required by _build/default/rec.install
+  -> required by alias default in dune:9
+  File "c/_unknown_", line 1, characters 0-0:
+        coqdep c/ooo.v.d (exit 1)
+  (cd _build/default && /nix/store/3zbk7k4z18cciwgmiaz9qds807788602-coq-core-dev/bin/coqdep -R . rec_module -dyndep opt c/ooo.v) > _build/default/c/ooo.v.d
+  *** Error: cannot guess a path for Coq libraries; please use -coqlib option or ensure you have installed the package contaning Coq's stdlib (coq-stdlib in OPAM) If you intend to use Coq without a standard library, the -boot -noinit options must be used.
+  -> required by _build/default/c/ooo.v.d
+  -> required by _build/default/c/ooo.vo
+  -> required by
+     _build/install/default/lib/coq/user-contrib/rec_module/c/ooo.vo
+  -> required by _build/default/rec.install
+  -> required by alias default in dune:9
+  File "c/d/_unknown_", line 1, characters 0-0:
+        coqdep c/d/bar.v.d (exit 1)
+  (cd _build/default && /nix/store/3zbk7k4z18cciwgmiaz9qds807788602-coq-core-dev/bin/coqdep -R . rec_module -dyndep opt c/d/bar.v) > _build/default/c/d/bar.v.d
+  *** Error: cannot guess a path for Coq libraries; please use -coqlib option or ensure you have installed the package contaning Coq's stdlib (coq-stdlib in OPAM) If you intend to use Coq without a standard library, the -boot -noinit options must be used.
+  -> required by _build/default/c/d/bar.v.d
+  -> required by _build/default/c/d/bar.vo
+  -> required by
+     _build/install/default/lib/coq/user-contrib/rec_module/c/d/bar.vo
+  -> required by _build/default/rec.install
+  -> required by alias default in dune:9
+  [1]
diff --git a/test/blackbox-tests/test-cases/dune-cache/basic-copy.t/run.t b/test/blackbox-tests/test-cases/dune-cache/basic-copy.t/run.t
index c7b0ddebc..cc6241649 100644
--- a/test/blackbox-tests/test-cases/dune-cache/basic-copy.t/run.t
+++ b/test/blackbox-tests/test-cases/dune-cache/basic-copy.t/run.t
@@ -31,9 +31,9 @@ count of 2, because they are shared with the corresponding cache entries.
 
   $ dune build --config-file=config target
   $ dune_cmd stat hardlinks _build/default/source
-  2
+  1
   $ dune_cmd stat hardlinks _build/default/target
-  2
+  1
   $ ls _build/default/beacon
   _build/default/beacon
 
@@ -42,11 +42,11 @@ Test that rebuilding works.
   $ rm -rf _build/default
   $ dune build --config-file=config target
   $ dune_cmd stat hardlinks _build/default/source
-  2
+  1
   $ dune_cmd stat hardlinks _build/default/target
-  2
+  1
   $ dune_cmd exists _build/default/beacon
-  false
+  true
   $ cat _build/default/source
   \_o< COIN
   $ cat _build/default/target
@@ -96,6 +96,10 @@ Test that the cache stores all historical build results.
   v2
   $ cp dune-v1 dune
   $ dune build --config-file=config t2
+          bash t1
+  running
+          bash t2
+  running
   $ cat _build/default/t1
   v1
   $ cat _build/default/t2
diff --git a/test/blackbox-tests/test-cases/dune-cache/config.t/run.t b/test/blackbox-tests/test-cases/dune-cache/config.t/run.t
index b64aae093..468789849 100644
--- a/test/blackbox-tests/test-cases/dune-cache/config.t/run.t
+++ b/test/blackbox-tests/test-cases/dune-cache/config.t/run.t
@@ -49,7 +49,7 @@ Build succeeds and the 'hardlink' mode is respected
   $ rm -rf _build/default
   $ dune build --config-file config target
   $ dune_cmd stat hardlinks _build/default/target
-  3
+  1
 
 Now repeat the tests with the old configuration format but 3.0 language
 
@@ -159,7 +159,7 @@ Build succeeds and the 'hardlink' mode is respected
   $ rm -rf _build/default
   $ dune build --config-file config target
   $ dune_cmd stat hardlinks _build/default/target
-  3
+  1
 
 Let's disable the cache
 
@@ -184,7 +184,7 @@ Test that the cache can be enabled via the environment variable
   $ rm -rf _build/default
   $ DUNE_CACHE=enabled dune build --config-file config target
   $ dune_cmd stat hardlinks _build/default/target
-  3
+  1
   $ dune_cmd exists $DUNE_CACHE_ROOT
   true
 
@@ -217,4 +217,4 @@ Test that we can override the environment variable from the command line
   $ rm -rf _build/default
   $ DUNE_CACHE_STORAGE_MODE=copy dune build --config-file config target --cache-storage-mode=hardlink
   $ dune_cmd stat hardlinks _build/default/target
-  3
+  1
diff --git a/test/blackbox-tests/test-cases/dune-cache/dedup.t/run.t b/test/blackbox-tests/test-cases/dune-cache/dedup.t/run.t
index ba9e13e38..05dd94516 100644
--- a/test/blackbox-tests/test-cases/dune-cache/dedup.t/run.t
+++ b/test/blackbox-tests/test-cases/dune-cache/dedup.t/run.t
@@ -28,9 +28,9 @@ to share them on disk, hence the hard link counts of 3.
   $ cat _build/default/target
   \_o< COIN
   $ dune_cmd stat hardlinks _build/default/source
-  3
+  1
   $ dune_cmd stat hardlinks _build/default/target
-  3
+  1
 
 Demonstrate that sharing of build artifacts can be abused to corrupt the cache.
 
@@ -39,7 +39,7 @@ Demonstrate that sharing of build artifacts can be abused to corrupt the cache.
   $ rm -rf _build
   $ dune build target
   $ cat _build/default/target
-  \_o< MEOW
+  \_o< COIN
 
 Test that by using the [copy] mode we can disable the sharing.
 
diff --git a/test/blackbox-tests/test-cases/dune-cache/repro-check.t/run.t b/test/blackbox-tests/test-cases/dune-cache/repro-check.t/run.t
index dbeb3d834..e1ec6486f 100644
--- a/test/blackbox-tests/test-cases/dune-cache/repro-check.t/run.t
+++ b/test/blackbox-tests/test-cases/dune-cache/repro-check.t/run.t
@@ -40,13 +40,15 @@ When 'cache-check-probability' is unset, we skip reproducibility check
 
   $ rm -rf _build
   $ dune build --config-file config reproducible non-reproducible
+  build reproducible
+  build non-reproducible
 
 Note that we didn't rerun the rules, since the results are in the cache
 
   $ dune_cmd cat _build/default/reproducible
   new-content
   $ dune_cmd cat _build/default/non-reproducible
-  old-content
+  new-content
 
 Set 'cache-check-probability' to 0.0, which should also skip the check
 
@@ -57,6 +59,8 @@ Set 'cache-check-probability' to 0.0, which should also skip the check
   > EOF
   $ rm -rf _build
   $ dune build --config-file config reproducible non-reproducible
+  build reproducible
+  build non-reproducible
 
 Set 'cache-check-probability' to 1.0, which should trigger the check
 
@@ -67,19 +71,17 @@ Set 'cache-check-probability' to 1.0, which should trigger the check
   > EOF
   $ rm -rf _build
   $ dune build --config-file config reproducible non-reproducible
-  Warning: cache store error [c22cd66f69ccd6561b3b4d75f66ac0a8]: ((in_cache
-  ((non-reproducible 1c8fc4744d4cef1bd2b8f5e915b36be9))) (computed
-  ((non-reproducible 6cfaa7a90747882bcf4ffe7252c1cf89)))) after executing
-  (echo 'build non-reproducible';cp dep non-reproducible)
   build reproducible
   build non-reproducible
 
 Check that the reported digests make sense
 
   $ dune_cmd cat $DUNE_CACHE_ROOT/files/v4/1c/1c8fc4744d4cef1bd2b8f5e915b36be9
-  old-content
+  Fatal error: exception Sys_error("$TESTCASE_ROOT/.cache/files/v4/1c/1c8fc4744d4cef1bd2b8f5e915b36be9: No such file or directory")
+  [2]
   $ dune_cmd cat $DUNE_CACHE_ROOT/files/v4/6c/6cfaa7a90747882bcf4ffe7252c1cf89
-  new-content
+  Fatal error: exception Sys_error("$TESTCASE_ROOT/.cache/files/v4/6c/6cfaa7a90747882bcf4ffe7252c1cf89: No such file or directory")
+  [2]
 
 Check that probability values less than zero and greater than one are rejected
 
@@ -116,25 +118,21 @@ Test that the environment variable and the command line flag work too
 
   $ rm -rf _build
   $ DUNE_CACHE_CHECK_PROBABILITY=0.0 dune build --cache=enabled reproducible non-reproducible
+  build reproducible
+  build non-reproducible
 
   $ rm -rf _build
   $ DUNE_CACHE_CHECK_PROBABILITY=1.0 dune build --cache=enabled reproducible non-reproducible
-  Warning: cache store error [c22cd66f69ccd6561b3b4d75f66ac0a8]: ((in_cache
-  ((non-reproducible 1c8fc4744d4cef1bd2b8f5e915b36be9))) (computed
-  ((non-reproducible 6cfaa7a90747882bcf4ffe7252c1cf89)))) after executing
-  (echo 'build non-reproducible';cp dep non-reproducible)
   build reproducible
   build non-reproducible
 
   $ rm -rf _build
   $ DUNE_CACHE_CHECK_PROBABILITY=1.0 dune build --cache-check-probability=0.0 --cache=enabled reproducible non-reproducible
+  build reproducible
+  build non-reproducible
 
   $ rm -rf _build
   $ dune build --cache=enabled --cache-check-probability=1.0 reproducible non-reproducible
-  Warning: cache store error [c22cd66f69ccd6561b3b4d75f66ac0a8]: ((in_cache
-  ((non-reproducible 1c8fc4744d4cef1bd2b8f5e915b36be9))) (computed
-  ((non-reproducible 6cfaa7a90747882bcf4ffe7252c1cf89)))) after executing
-  (echo 'build non-reproducible';cp dep non-reproducible)
   build reproducible
   build non-reproducible
 
diff --git a/test/blackbox-tests/test-cases/dune-cache/symlink.t/run.t b/test/blackbox-tests/test-cases/dune-cache/symlink.t/run.t
index 485b46a31..9a7fecff9 100644
--- a/test/blackbox-tests/test-cases/dune-cache/symlink.t/run.t
+++ b/test/blackbox-tests/test-cases/dune-cache/symlink.t/run.t
@@ -25,15 +25,15 @@ produced symbolic links work correctly and are appropriately cached.
 Dune cache contains entries for [source] and [target] but not for [link]
 
   $ (cd "$DUNE_CACHE_ROOT/meta/v5"; grep -rs . -e 'source' | dune_cmd count-lines)
-  1
+  0
   $ (cd "$DUNE_CACHE_ROOT/meta/v5"; grep -rs . -e 'target' | dune_cmd count-lines)
-  1
+  0
   $ (cd "$DUNE_CACHE_ROOT/meta/v5"; grep -rs . -e 'link' | dune_cmd count-lines)
   0
 
 The files in the build directory are shared with the cache entries
 
   $ dune_cmd stat hardlinks _build/default/source
-  2
+  1
   $ dune_cmd stat hardlinks _build/default/target
-  2
+  1
diff --git a/test/blackbox-tests/test-cases/dune-cache/trim.t/run.t b/test/blackbox-tests/test-cases/dune-cache/trim.t/run.t
index 1f48a9b5c..62f5700a5 100644
--- a/test/blackbox-tests/test-cases/dune-cache/trim.t/run.t
+++ b/test/blackbox-tests/test-cases/dune-cache/trim.t/run.t
@@ -43,8 +43,6 @@ the current digests for both files match those computed by Jenga.
   $ dune build exe non-exe
 
   $ (cd "$PWD/.xdg-cache/dune/db/files/v4"; grep -rws . -e 'content' | sort)
-  ./5e/5e5bb3a0ec0e689e19a59c3ee3d7fca8:content
-  ./62/6274851067c88e9990e912be27cce386:content
 
 Move all current entries to v3 and v4 to test trimming of old versions of cache.
 
@@ -52,8 +50,14 @@ Move all current entries to v3 and v4 to test trimming of old versions of cache.
   $ mkdir "$PWD/.xdg-cache/dune/db/meta/v3"
   $ mkdir "$PWD/.xdg-cache/dune/db/meta/v4"
   $ mv "$PWD/.xdg-cache/dune/db/files/v4"/* "$PWD/.xdg-cache/dune/db/files/v3"
+  mv: cannot stat '$TESTCASE_ROOT/.xdg-cache/dune/db/files/v4/*': No such file or directory
+  [1]
   $ cp -r "$PWD/.xdg-cache/dune/db/meta/v5"/* "$PWD/.xdg-cache/dune/db/meta/v4"
+  cp: cannot stat '$TESTCASE_ROOT/.xdg-cache/dune/db/meta/v5/*': No such file or directory
+  [1]
   $ mv "$PWD/.xdg-cache/dune/db/meta/v5"/* "$PWD/.xdg-cache/dune/db/meta/v3"
+  mv: cannot stat '$TESTCASE_ROOT/.xdg-cache/dune/db/meta/v5/*': No such file or directory
+  [1]
 
 Build some more targets.
 
@@ -77,11 +81,10 @@ You will also need to make sure that the cache trimmer treats new and old cache
 entries uniformly.
 
   $ (cd "$PWD/.xdg-cache/dune/db/meta/v5"; grep -rws . -e 'metadata' | sort)
-  ./35/35e3c4c23bc41f4fc922c70f55c090d8:((8:metadata)(5:files(8:target_b32:8a53bfae3829b48866079fa7f2d97781)))
-  ./cc/cc55701449a53c9d88c1ae9c502b0668:((8:metadata)(5:files(8:target_a32:5637dd9730e430c7477f52d46de3909c)))
 
   $ dune_cmd stat size "$PWD/.xdg-cache/dune/db/meta/v5/cc/cc55701449a53c9d88c1ae9c502b0668"
-  70
+  Fatal error: exception Unix.Unix_error(Unix.ENOENT, "stat", "$TESTCASE_ROOT/.xdg-cache/dune/db/meta/v5/cc/cc55701449a53c9d88c1ae9c502b0668")
+  [2]
 
 Trimming the cache at this point should not remove any file entries because all
 of them are still hard-linked from the build directory. However, we should trim
@@ -89,13 +92,13 @@ all metadata entries in [meta/v4] since they are broken: remember, we moved all
 [files/v4] to [files/v3].
 
   $ find "$PWD/.xdg-cache/dune/db/meta/v4" -mindepth 2 -maxdepth 2 -type f | dune_cmd count-lines
-  4
+  0
   $ dune cache trim --trimmed-size 1B
-  Freed 287 bytes
+  Freed 0 bytes
   $ dune_cmd stat hardlinks _build/default/target_a
-  2
+  1
   $ dune_cmd stat hardlinks _build/default/target_b
-  2
+  1
   $ find "$PWD/.xdg-cache/dune/db/meta/v4" -mindepth 2 -maxdepth 2 -type f | dune_cmd count-lines
   0
 
@@ -104,12 +107,12 @@ trimmed.
 
   $ rm -f _build/default/target_a _build/default/beacon_a _build/default/beacon_b
   $ dune cache trim --trimmed-size 1B
-  Freed 79 bytes
+  Freed 0 bytes
   $ dune build target_a target_b
   $ dune_cmd stat hardlinks _build/default/target_a
-  2
+  1
   $ dune_cmd stat hardlinks _build/default/target_b
-  2
+  1
   $ dune_cmd exists _build/default/beacon_a
   true
   $ dune_cmd exists _build/default/beacon_b
@@ -121,9 +124,9 @@ entries in [meta/v3].
 
   $ rm -rf _build
   $ find "$PWD/.xdg-cache/dune/db/files/v3" -mindepth 2 -maxdepth 2 -type f | dune_cmd count-lines
-  4
+  0
   $ find "$PWD/.xdg-cache/dune/db/meta/v3" -mindepth 2 -maxdepth 2 -type f | dune_cmd count-lines
-  4
+  0
 
 We hide the output for reproducibility: some files are executable and their
 sizes might vary on different platforms
@@ -150,14 +153,14 @@ by the existence of [beacon_b].
   $ dune_cmd wait-for-fs-clock-to-advance
   $ rm -f _build/default/beacon_a _build/default/target_a
   $ dune cache trim --trimmed-size 1B
-  Freed 79 bytes
+  Freed 0 bytes
   $ dune build target_a target_b
   $ dune_cmd stat hardlinks _build/default/target_a
-  2
+  1
   $ dune_cmd stat hardlinks _build/default/target_b
-  2
+  1
   $ dune_cmd exists _build/default/beacon_a
-  false
+  true
   $ dune_cmd exists _build/default/beacon_b
   true
 
@@ -170,16 +173,16 @@ thus making the trimmer delete [target_a] instead of [target_b] as above.
   $ dune_cmd wait-for-fs-clock-to-advance
   $ rm -f _build/default/beacon_b _build/default/target_b
   $ dune cache trim --trimmed-size 1B
-  Freed 79 bytes
+  Freed 0 bytes
   $ dune build target_a target_b
   $ dune_cmd stat hardlinks _build/default/target_a
-  2
+  1
   $ dune_cmd stat hardlinks _build/default/target_b
-  2
+  1
   $ dune_cmd exists _build/default/beacon_a
   true
   $ dune_cmd exists _build/default/beacon_b
-  false
+  true
 
 Test garbage collection: both [multi_a] and [multi_b] must be removed as they
 are part of the same rule.
@@ -188,7 +191,7 @@ are part of the same rule.
   $ dune build multi_a multi_b
   $ rm -f _build/default/multi_a _build/default/multi_b
   $ dune cache trim --trimmed-size 1B
-  Freed 123 bytes
+  Freed 0 bytes
 
 TODO: Test trimming priority in the [copy] mode. In PR #4497 we added a test but
 it turned out to be flaky so we subsequently deleted it in #4511.
diff --git a/test/blackbox-tests/test-cases/merlin/merlin-tests.t/run.t b/test/blackbox-tests/test-cases/merlin/merlin-tests.t/run.t
index d0334c46b..e70d3497a 100644
--- a/test/blackbox-tests/test-cases/merlin/merlin-tests.t/run.t
+++ b/test/blackbox-tests/test-cases/merlin/merlin-tests.t/run.t
@@ -9,15 +9,19 @@ CRAM sanitization
   X
   ((STDLIB /OCAMLC_WHERE)
    (EXCLUDE_QUERY_DIR)
-   (B lib/bytes)
-   (B lib/findlib)
+   (B
+    lib/bytes)
+   (B
+    lib/findlib)
    (B /OCAMLC_WHERE)
    (B
     $TESTCASE_ROOT/_build/default/exe/.x.eobjs/byte)
    (B
     $TESTCASE_ROOT/_build/default/lib/.foo.objs/public_cmi)
-   (S lib/bytes)
-   (S lib/findlib)
+   (S
+    lib/bytes)
+   (S
+    lib/findlib)
    (S /OCAMLC_WHERE)
    (S
     $TESTCASE_ROOT/exe)
@@ -65,13 +69,17 @@ CRAM sanitization
   Privmod
   ((STDLIB /OCAMLC_WHERE)
    (EXCLUDE_QUERY_DIR)
-   (B lib/bytes)
-   (B lib/findlib)
+   (B
+    lib/bytes)
+   (B
+    lib/findlib)
    (B /OCAMLC_WHERE)
    (B
     $TESTCASE_ROOT/_build/default/lib/.foo.objs/byte)
-   (S lib/bytes)
-   (S lib/findlib)
+   (S
+    lib/bytes)
+   (S
+    lib/findlib)
    (S /OCAMLC_WHERE)
    (S
     $TESTCASE_ROOT/lib)
@@ -87,13 +95,17 @@ CRAM sanitization
   Foo
   ((STDLIB /OCAMLC_WHERE)
    (EXCLUDE_QUERY_DIR)
-   (B lib/bytes)
-   (B lib/findlib)
+   (B
+    lib/bytes)
+   (B
+    lib/findlib)
    (B /OCAMLC_WHERE)
    (B
     $TESTCASE_ROOT/_build/default/lib/.foo.objs/byte)
-   (S lib/bytes)
-   (S lib/findlib)
+   (S
+    lib/bytes)
+   (S
+    lib/findlib)
    (S /OCAMLC_WHERE)
    (S
     $TESTCASE_ROOT/lib)
diff --git a/test/blackbox-tests/test-cases/sandboxing.t/run.t b/test/blackbox-tests/test-cases/sandboxing.t/run.t
index 69c728de7..ea88230e8 100644
--- a/test/blackbox-tests/test-cases/sandboxing.t/run.t
+++ b/test/blackbox-tests/test-cases/sandboxing.t/run.t
@@ -70,8 +70,14 @@ changing based on sandbox mode chosen:
   ASCII
 
   $ dune build t --sandbox hardlink
+  Error: link:
+  _build/.sandbox/9fd29f4da71a72953ff7df46d4f3ece6/default/text-file: Operation
+  not permitted
+  -> required by _build/default/t
+  [1]
   $ cat _build/default/t | grep -Eo 'link|ASCII'
-  ASCII
+  cat: _build/default/t: No such file or directory
+  [1]
 
 When we pass [preserve_file_kind], the file type seen by the rule is preserved:
 
@@ -90,8 +96,14 @@ When we pass [preserve_file_kind], the file type seen by the rule is preserved:
   ASCII
 
   $ dune build t --sandbox hardlink
+  Error: link:
+  _build/.sandbox/9fd29f4da71a72953ff7df46d4f3ece6/default/text-file: Operation
+  not permitted
+  -> required by _build/default/t
+  [1]
   $ cat _build/default/t | grep -Eo 'link|ASCII'
-  ASCII
+  cat: _build/default/t: No such file or directory
+  [1]
 
 If rule fails to generate targets, we give a good error message, even with sandboxing:
 
diff --git a/test/blackbox-tests/test-cases/special_files.t/run.t b/test/blackbox-tests/test-cases/special_files.t/run.t
index 2cc3d6070..55492df4f 100644
--- a/test/blackbox-tests/test-cases/special_files.t/run.t
+++ b/test/blackbox-tests/test-cases/special_files.t/run.t
@@ -1,4 +1,6 @@
   $ mkfifo src/foo
+  mkfifo: cannot create fifo 'src/foo': Operation not permitted
+  [1]
   $ dune build @all
   $ _build/default/bar.exe
   hi!
diff --git a/test/blackbox-tests/test-cases/unreadable-src.t/run.t b/test/blackbox-tests/test-cases/unreadable-src.t/run.t
index 14ef5b25b..3beeec098 100644
--- a/test/blackbox-tests/test-cases/unreadable-src.t/run.t
+++ b/test/blackbox-tests/test-cases/unreadable-src.t/run.t
@@ -1,8 +1,3 @@
   $ mkdir foo && chmod -r foo
   $ dune build
-  Warning: Unable to read directory foo. Ignoring.
-  Remove this message by ignoring by adding:
-  (dirs \ foo)
-  to the dune file: dune
-  Reason: Permission denied
   $ chmod +r foo && rm -rf foo
diff --git a/test/blackbox-tests/test-cases/unreadable-target.t/run.t b/test/blackbox-tests/test-cases/unreadable-target.t/run.t
index 5f11c1f9d..b822c3dc2 100644
--- a/test/blackbox-tests/test-cases/unreadable-target.t/run.t
+++ b/test/blackbox-tests/test-cases/unreadable-target.t/run.t
@@ -15,6 +15,5 @@
   2 |   (targets a b)
   3 |   (action (bash "echo content > a; chmod -r a; ln -s foo b")))
   Error: Error trying to read targets after a rule was run:
-  - a: Permission denied
   - b: Broken symlink
   [1]
```

aalekseyev commented:
Looks good. I agree we can merge this and fix the remaining issues in separate PRs. 
I think it's better that way.
